web/gridt/resources/user.py

```diff
@@ -42,7 +53,54 @@ def post(self):
 
         if not current_identity.verify_password(res["old_password"]):
             return ({"message": "Failed to identify user with given password."}, 400)
-        
+
         if current_identity.verify_password(res["old_password"]):
             current_identity.hash_password(res["new_password"])
-            return ({"message": "Successfully changed password."}, 200)
\ No newline at end of file
+
+            send_password_change_notification(current_identity.email)
+            return ({"message": "Successfully changed password."}, 200)
+
+
+class RequestPasswordResetResource(Resource):
+    schema = RequestPasswordResetSchema()
+
+    def post(self):
+        try:
+            res = self.schema.load(request.get_json())
+        except ValidationError as error:
+            field = list(error.messages.keys())[0]
+            return ({"message": f"{field}: {error.messages[field][0]}"}, 400)
+
+        if not current_app.config["EMAIL_API_KEY"]:
+            return ({"message": "Could not send e-mail; API key not available."}, 500)
+
+        user = User.find_by_email(res["email"])
+        if not user:
+            # We don't want malicious users to know whether or not an e-mail is in our database.
+            return ({"message": "Recovery e-mail successfully sent."}, 200)
+
+        token = user.get_password_reset_token()
+        send_password_reset_email(user.email, token)
+        return ({"message": "Recovery e-mail successfully sent."}, 200)
+
+
+class ResetPasswordResource(Resource):
+    schema = ResetPasswordSchema()
+
+    def post(self):
+        secret_key = current_app.config["SECRET_KEY"]
+        try:
+            res = self.schema.load(request.get_json())
+        except ValidationError as error:
+            field = list(error.messages.keys())[0]
+            return ({"message": f"{field}: {error.messages[field][0]}"}, 400)
+
+        token_decoded = jwt.decode(res["token"], secret_key, algorithms=["HS256"])
+        user_id = token_decoded["user_id"]
+        password = res["password"]
+
+        user = User.find_by_id(user_id)
+        user.hash_password(password)
```

Comment:
This function should be renamed to hash_and_store_password or something like that, as it is not entirely clear from just reading what you are doing here.